Encode google API key with Base64

ayushiitr2901 opened the pull request:
This Pull Request is a continuation of the discussion on issue #9521. The basic aim was to encode the Google API key with Base64 format before parsing it into JSON, and then decoding it at the JS side.
- [ ] issues: fixes #9521 
- [ ] tested on local virtual env

ayushiitr2901 commented:
Apologies, I think I forgot to update related test cases. I shall correct it commit again.

mattpap commented on bokeh/util/serialization.py:
```diff
@@ -569,6 +570,18 @@ def decode_base64_dict(data):
+        string
+
+    '''
+    b64_string = base64.b64encode(value.encode("ascii"))
```
Why limit strings to ASCII? `utf-8` encoding seems to be a much better choice in my opinion.

ayushiitr2901 replied:
Surely utf-8 is better than ASCII, I will change it.

bryevdv replied:
Is this extra function even necessary? I would not be opposed to just calling 
```
base64.b64encode(value.encode("utf-8"))
``` 
directly in `Base64String` and avoiding adding new functions.

ayushiitr2901 replied:
I think this sounds better, I shall remove this and define base64 directly into `Base64String`.

mattpap commented on bokeh/plotting/gmap.py:
```diff
@@ -230,8 +230,9 @@ def gmap(google_api_key, map_options, **kwargs):
-
-    return GMap(api_key=google_api_key, map_options=map_options, **kwargs)
+    encoded_key = Base64String()
+    encoded_key = encoded_key.serialize_value(google_api_key)
```
This seems completely unnecessary. It's up to the properties' system to call this method when needed. If I follow the logic correctly, this will result in the API key being encoded twice.

ayushiitr2901 replied:
I have defined the Base64Encode class to variable 'api_key'. Will passing 'google_api_key' directly to it without explicitly using _'encoded_key.serialize_value(google_api_key)'_ encode the string?

bryevdv replied:
@ayushiitr2901 The intention behind all this is that the key can be specified normally on the Python side:
```
return GMap(api_key=google_api_key, map_options=map_options, **kwargs)
```
And that the property system will automatically, transparently encode the value before serializing. This kind of step should not be required of users.

bryevdv replied:
Or perhaps a clearer way of putting things: In the making of this PR, *no existing example code should need to be changed in any way*

bryevdv commented on bokeh/models/map_plots.py:
```diff
@@ -21,7 +21,7 @@
 from ..core.enums import MapType
 from ..core.has_props import abstract
-from ..core.properties import JSON, Bool, Enum, Float, Instance, Int, Override, String
+from ..core.properties import JSON, Bool, Enum, Float, Instance, Int, Override, Base64String
```
imports must be alphabetized

bryevdv commented on bokeh/plotting/gmap.py:
```diff
@@ -17,7 +17,7 @@
 # Bokeh imports
 from ..core.enums import HorizontalLocation, VerticalLocation
-from ..core.properties import Auto, Either, Enum, Instance, Int, Seq, String
+from ..core.properties import Auto, Either, Enum, Instance, Int, Seq, String, Base64String
```
imports must be alphabetized

bryevdv commented on bokeh/core/properties.py:
```diff
@@ -302,6 +303,7 @@ class SomeModel(Model):
 from .property.numeric import Percent; Percent
 from .property.numeric import PositiveInt; PositiveInt
 from .property.numeric import Size; Size
+from .property.numeric import Base64String; Base64String
```
imports must be aphabetized

bryevdv commented on bokeh/util/serialization.py:
```diff
@@ -569,6 +570,18 @@ def decode_base64_dict(data):
         array = array.reshape(data['shape'])
     return array
 
+def encode_base64_string(str, value):
```
The signature is wrong here, perhaps you mean:
```
def encode_base64_string(value: str) -> str
```

bryevdv commented on bokeh/util/serialization.py:
```diff
@@ -569,6 +570,18 @@ def decode_base64_dict(data):
+    ''' Encode a ascii string using Base64.
+
+    Args:
+        str : a string to encode
```
arg name is `value` (in general, should try to avoid using built-in Python names like `str` as variable names)

bryevdv commented on bokeh/core/property/numeric.py:
```diff
@@ -269,6 +271,14 @@ class Angle(Float):
     '''
     pass
 
+class Base64String(String):
```
This is in the wrong module, this should be in `bokeh/core/property/string.py`

ayushiitr2901 replied:
At the moment, I have moved the class from `numeric.py` to `regex.py`, but have not renamed the file. There might be various places where regex.py file is currently used and renaming it would cause a problem. How should I tackle this? Also could you please elaborate on,

> we would need to leave a deprecated regex.py as well, to maintain back compat

bryevdv replied:
@ayushiitr2901 Right I am suggesting renaming `regex.py` to `string.py` but then also adding a new `regex.py` that does nothin except import the `Regex` property (to preserve compatibility) and also calls `deprecated(...)` to note that the`regex` module is deprecated.

bryevdv commented on bokeh/core/property/regex.py:
```diff
@@ -90,6 +92,23 @@ def validate(self, value, detail=True):
+    def serialize_value(self, value):
+
+        if isinstance(value, str):
+            value = base64.b64encode(value.encode("utf-8"))
```
So this will need one more `.decode("utf-8") at the end so that the properties system can serialize the base64 value (plain bytes can't be serialized). 

there should not be any changes needed on the JS side.

bryevdv commented:
@ayushiitr2901 There is also an import formatting issue. The simplest way to resolve this will be to make sure you have all the dev dependencies installed, *including* a `pip install channels` and then to run `isort -rc bokeh` at the top level of the repo. Otherwise you can also look in the failed codebase build job to see a diff that shows the expected format.

ayushiitr2901 commented:
I have renamed `regex.py` to `string.py` as suggested and updated and renamed the related test cases as well. Also, I have added a Deprecation message for module `regex.`

ayushiitr2901 commented:
Test for module `string.py` shall be named `test_string.py`, but a test name the same already exists. Thus, I have renamed the test for `string.py`.

bryevdv commented:
> Test for module string.py shall be named test_string.py, but a test name the same already exists. Thus, I have renamed the test for string.py.

I don't see at `test_string.py` in the same directory? I think it should be named `test_string.py` all of the tests in that directory (and mostly generally) follow the pattern `test_XYX.py` here XYZ is the module name of the module under tests. We should keep to that here.

bryevdv commented on bokeh/core/property/string.py:
```diff
@@ -0,0 +1,122 @@
+
+    def serialize_value(self, value):
+
+        if isinstance(value, str):
```
is this `isinstance` necessary? AFAIK the validation on the parent class `String` will raise an error for anything that is not a `str`

bryevdv replied:
Ignore this :) I forgot the value is also nullable

mattpap replied:
In that case it would make much more sense to use `value is not None` (i.e. exclude a single value). Also it would be helpful if all new code was statically typed (at least arguments and return values), then there would be no room for confusion.

bryevdv replied:
They are equivalent from a readability POV to me, but `is not None` is microscopically faster (10s of ns) so switching to that is fine with me.

bryevdv replied:
@mattpap if we want new contributors to add type annotations in places they are not already, then we would really need a dev guide section on how to do that to point to

bryevdv commented:
I think after the test file rename this will be ready to merge!

mattpap commented on bokeh/core/property/string.py:
```diff
@@ -0,0 +1,122 @@
+        value : a string to encode
+
+    Returns:
+        string
```
This docstring is misplaced. Perhaps was supposed do be added to `serialize_value()` method.

ayushiitr2901 replied:
Oh, I forgot to update that. I will change it.

mattpap commented on bokeh/core/property/regex.py:
```diff
@@ -34,62 +32,12 @@
     'Regex',
 )
 
+deprecated("bokeh/core/property/regex is deprecated, use bokeh/core/property/string module instead")
```
I would be better to use dots instead of slashes to indicate modules. Otherwise add `.py` at the end of those paths.

mattpap replied:
Also, preferably use the three argument variant of `deprecated()` with `version=(2, 0, 1)`.

ayushiitr2901 commented:
Again, it shows an error if I try to name the test file the same as test_string.py,
```
___________ ERROR collecting tests/unit/bokeh/util/test_string.py _____________
import file mismatch:
imported module 'test_string' has this __file__ attribute:
  /home/runner/work/bokeh/bokeh/tests/unit/bokeh/core/property/test_string.py
which is not the same as the test file we want to collect:
  /home/runner/work/bokeh/bokeh/tests/unit/bokeh/util/test_string.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

bryevdv commented:
@ayushiitr2901 Hrm, well that is certainly an annoying limitation of the test runner. In that case please name it `test_string_properties.py`

bryevdv commented:
@ayushiitr2901 where did the Typescript updates go? Sorry, in my last comment when I said no changes were necessary with the extra encode, I meant: no further change, from the current state. We definitely need the  call to `atob`

ayushiitr2901 commented:
I have been thinking that removing the` atob` function should be wrong. I will fix it.
@bryevdv What exactly is the effect of adding a .decode("utf-8") at the last?

bryevdv commented:
>  What exactly is the effect of adding a .decode("utf-8") at the last?

There is no Bokeh property that can serialize raw bytes, so that converts it to a string type, which can be

bryevdv commented:
tested locally, will merge when green, thanks @ayushiitr2901
